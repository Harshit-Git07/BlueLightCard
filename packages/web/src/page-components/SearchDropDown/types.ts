export type SearchDropDownProps = {
  onSearchCompanyChange: (companyId: string, company: string) => void;
  onSearchCategoryChange: (categoryId: string, categoryName: string) => void;
};
