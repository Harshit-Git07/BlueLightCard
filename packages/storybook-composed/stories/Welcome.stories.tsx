import type { Meta, StoryObj } from '@storybook/react';
import WelcomeComponent from './Welcome';

const meta: Meta<typeof WelcomeComponent> = {
  component: WelcomeComponent,
};

export default meta;

type Story = StoryObj<typeof WelcomeComponent>;

export const Welcome: Story = {
  args: {},
  argTypes: {},
};
