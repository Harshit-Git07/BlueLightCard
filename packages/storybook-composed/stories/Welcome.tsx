import React, { FC } from 'react';

const WelcomeComponent: FC = () => (
  <div>
    <p>
      Welcome to the Storybook Composed instance. Here you can find all of our stories across Web,
      Mobile Hybrid, and Shared UI for every brand. Please view these stories from the sidebar.
    </p>
  </div>
);

export default WelcomeComponent;
