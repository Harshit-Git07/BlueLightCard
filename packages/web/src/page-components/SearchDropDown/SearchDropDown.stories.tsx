import { Meta, StoryFn } from '@storybook/react';
import { SearchDropDownPresenter } from './SearchDropDown';

const componentMeta: Meta<typeof SearchDropDownPresenter> = {
  title: 'Components/SearchDropDown',
  component: SearchDropDownPresenter,
  argTypes: {},
  parameters: {
    controls: {
      exclude: [
        'categories',
        'companies',
        'isLoading',
        'onSearchCategoryChange',
        'onSearchCompanyChange',
      ],
    },
  },
};

const mockData = {
  data: {
    response: {
      categories: [
        {
          id: '11',
          name: 'Children and toys',
          __typename: 'CategoryMenu',
        },
        {
          id: '3',
          name: 'Days out',
          __typename: 'CategoryMenu',
        },
        {
          id: '1',
          name: 'Electrical and phones',
          __typename: 'CategoryMenu',
        },
        {
          id: '6',
          name: 'Entertainment',
          __typename: 'CategoryMenu',
        },
        {
          id: '8',
          name: 'Fashion',
          __typename: 'CategoryMenu',
        },
        {
          id: '16',
          name: 'Featured',
          __typename: 'CategoryMenu',
        },
        {
          id: '9',
          name: 'Financial and insurance',
          __typename: 'CategoryMenu',
        },
        {
          id: '15',
          name: 'Food and drink',
          __typename: 'CategoryMenu',
        },
        {
          id: '5',
          name: 'Gifts',
          __typename: 'CategoryMenu',
        },
        {
          id: '4',
          name: 'Health and beauty',
          __typename: 'CategoryMenu',
        },
        {
          id: '2',
          name: 'Holidays and travel',
          __typename: 'CategoryMenu',
        },
        {
          id: '10',
          name: 'Home and garden',
          __typename: 'CategoryMenu',
        },
        {
          id: '7',
          name: 'Motoring',
          __typename: 'CategoryMenu',
        },
      ],
      companies: [
        {
          id: '26529',
          name: ' Youth & Earth',
          __typename: 'CompanyMenu',
        },
        {
          id: '14779',
          name: '1 Stop Barber Shop ',
          __typename: 'CompanyMenu',
        },
        {
          id: '15939',
          name: '11 Degrees',
          __typename: 'CompanyMenu',
        },
        {
          id: '23876',
          name: '18montrose',
          __typename: 'CompanyMenu',
        },
        {
          id: '20128',
          name: '1st 4 Signs Ltd',
          __typename: 'CompanyMenu',
        },
        {
          id: '13571',
          name: '360 Play',
          __typename: 'CompanyMenu',
        },
        {
          id: '5961',
          name: '361 Property Management ',
          __typename: 'CompanyMenu',
        },
        {
          id: '15943',
          name: '3RD ROCK',
          __typename: 'CompanyMenu',
        },
        {
          id: '19750',
          name: '43T Clothing',
          __typename: 'CompanyMenu',
        },
      ],
    },
  },
};

const DefaultTemplate: StoryFn<typeof SearchDropDownPresenter> = (args) => (
  <SearchDropDownPresenter {...args}></SearchDropDownPresenter>
);

export const Default = DefaultTemplate.bind({});

Default.args = {
  categories: mockData.data.response.categories,
  companies: mockData.data.response.companies,
  isLoading: false,
  onSearchCategoryChange: () => {},
  onSearchCompanyChange: () => {},
};

export default componentMeta;
