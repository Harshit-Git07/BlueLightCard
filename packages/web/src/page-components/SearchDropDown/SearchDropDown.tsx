import React, { FC, useContext } from 'react';
import { SearchDropDownProps } from './types';
import Link from '@/components/Link/Link';
import InputTypeSelect from '@/components/InputTypeSelect/InputTypeSelect';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faLocationDot } from '@fortawesome/pro-solid-svg-icons';
import { PillGroup } from '@bluelightcard/shared-ui';
import UserContext from '@/context/User/UserContext';
import useFetchCompaniesOrCategories, {
  CategoryType,
  CompanyType,
} from '@/hooks/useFetchCompaniesOrCategories';

type SearchDropDownPresenterProps = {
  onSearchCategoryChange: (categoryId: string, categoryName: string) => void;
  onSearchCompanyChange: (companyId: string, companyName: string) => void;
  categories: CategoryType[];
  companies: CompanyType[];
  isLoading: boolean;
};

export const SearchDropDownPresenter = ({
  categories,
  companies,
  isLoading,
  onSearchCategoryChange,
  onSearchCompanyChange,
}: SearchDropDownPresenterProps) => {
  const changeCategoryHandler = (categoryId: string, categoryName: string) => {
    onSearchCategoryChange(categoryId, categoryName);
  };

  const changeCompanyHandler = (companyId: string, companyName: string) => {
    onSearchCompanyChange(companyId, companyName);
  };

  return (
    <div className="mx-auto pt-[54px] pb-[100px] max-w-[1408px]">
      <div className="flex gap-2 px-4 py-[10px]">
        <Link href={`/nearme.php`} useLegacyRouting>
          <FontAwesomeIcon icon={faLocationDot} size="lg" className="mr-2" />
          <span className="font-typography-body font-typography-body-weight text-typography-body leading-typography-body tracking-typography-body">
            Offers near you
          </span>
        </Link>
      </div>
      <div className="my-[22px] px-4">
        <PillGroup
          title={'Browse categories'}
          pillGroup={categories.map((cat) => ({
            id: Number(cat.id),
            label: cat.name,
          }))}
          onSelectedPill={(id) => {
            const selectedCategory = categories.find((cat) => cat.id === id.toString());
            if (selectedCategory) {
              changeCategoryHandler(selectedCategory.id, selectedCategory.name);
            }
          }}
        />
      </div>
      <div className="px-4">
        <p className="mb-2 font-typography-title-large font-typography-title-large-weight text-typography-title-large leading-typography-title-large tracking-typography-title-large text-colour-onSurface-light dark:text-colour-onSurface-dark">
          Browse companies
        </p>
        <InputTypeSelect
          options={companies.map((cat) => ({
            id: cat.id,
            label: cat.name,
          }))}
          placeholder={'Select company'}
          onSelect={(id: string) => {
            const selectedCategory = companies.find((comp) => comp.id === id);
            if (selectedCategory) {
              changeCompanyHandler(selectedCategory.id, selectedCategory.name);
            }
          }}
        />
      </div>
    </div>
  );
};

const SearchDropDown: FC<SearchDropDownProps> = ({
  onSearchCategoryChange,
  onSearchCompanyChange,
}) => {
  const userCtx = useContext(UserContext);
  const { categories, companies, isLoading } = useFetchCompaniesOrCategories(userCtx);

  return (
    <SearchDropDownPresenter
      isLoading={isLoading}
      companies={companies}
      categories={categories}
      onSearchCategoryChange={onSearchCategoryChange}
      onSearchCompanyChange={onSearchCompanyChange}
    />
  );
};

export default SearchDropDown;
