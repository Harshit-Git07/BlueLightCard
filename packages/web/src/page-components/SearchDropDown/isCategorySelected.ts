export const isCategorySelected = (categoryId: string, pathname: string): boolean => {
  if (!pathname.startsWith('/offers.php?cat=true&type=')) {
    return false;
  }
  return pathname.includes(categoryId);
};
