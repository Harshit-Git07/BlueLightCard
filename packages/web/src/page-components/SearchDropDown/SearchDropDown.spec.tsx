import { render, fireEvent, within, cleanup } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import { composeStories } from '@storybook/react';
import * as stories from './SearchDropDown.stories';
import '@testing-library/jest-dom';

const { Default, Interaction } = composeStories(stories);

import { act } from 'react';

jest.mock('@/hooks/useFetchCompaniesOrCategories');

describe('SearchDropDown Component', () => {
  beforeEach(() => {
    window.HTMLElement.prototype.scrollIntoView = jest.fn();
  });

  afterEach(() => {
    cleanup();
    document.body.classList.remove('no-scroll');
  });

  describe('Basic Rendering', () => {
    it('renders the dropdown correctly', () => {
      const { container } = render(<Default />);
      expect(container).toMatchSnapshot();
    });
  });

  describe('Interactivity', () => {
    it('handles user interactions correctly', async () => {
      const onSearchCategoryChange = jest.fn();
      const onSearchCompanyChange = jest.fn();
      const onClose = jest.fn();

      const { container } = render(
        <Interaction
          onSearchCategoryChange={onSearchCategoryChange}
          onSearchCompanyChange={onSearchCompanyChange}
          onClose={onClose}
        />
      );

      const category = within(container).getByText('Children and toys');
      fireEvent.click(category);

      const companySearch = within(container).getByRole('combobox', {
        name: 'Search for a company',
        hidden: true,
      });
      await act(() => userEvent.type(companySearch, 'you'));

      const company = await within(container).findByText('Youth & Earth');
      fireEvent.click(company);

      // Close dropdown by clicking the overlay
      const overlay = within(container).getByTestId('search-dropdown-overlay');
      fireEvent.click(overlay);

      expect(onSearchCategoryChange).toHaveBeenCalledWith('11', 'Children and toys');
      expect(onSearchCompanyChange).toHaveBeenCalledWith('26529', 'Youth & Earth');
      expect(onClose).toHaveBeenCalled();
    });

    it('closes the dropdown when a category is clicked', () => {
      const onSearchCategoryChange = jest.fn();

      const { container } = render(
        <Interaction
          isOpen={true}
          onSearchCategoryChange={onSearchCategoryChange}
          onSearchCompanyChange={jest.fn()}
          onClose={jest.fn()}
        />
      );

      const category = within(container).getByText('Children and toys');
      fireEvent.click(category);

      expect(onSearchCategoryChange).toHaveBeenCalledWith('11', 'Children and toys');
      const optionsListbox = container.querySelector('[role="listbox"]');
      expect(optionsListbox).not.toBeInTheDocument();
    });

    it('closes the dropdown when a company is clicked', async () => {
      const onSearchCompanyChange = jest.fn();

      const { container } = render(
        <Interaction
          onSearchCategoryChange={jest.fn()}
          onSearchCompanyChange={onSearchCompanyChange}
          onClose={jest.fn()}
        />
      );

      const companySearch = within(container).getByRole('combobox', {
        name: 'Search for a company',
        hidden: true,
      });

      await act(() => userEvent.type(companySearch, 'you'));
      const optionsListbox = container.querySelector('[role="listbox"]');
      expect(optionsListbox).toBeInTheDocument();

      const company = await within(container).findByText('Youth & Earth');
      fireEvent.click(company);

      expect(onSearchCompanyChange).toHaveBeenCalledWith('26529', 'Youth & Earth');
      expect(optionsListbox).not.toBeInTheDocument();
    });
  });

  describe('Scroll Behavior', () => {
    const onSearchCategoryChange = jest.fn();
    const onSearchCompanyChange = jest.fn();
    const onClose = jest.fn();

    it('adds the no-scroll class to the body when isOpen is true', () => {
      render(
        <Default
          isOpen={true}
          onSearchCategoryChange={onSearchCategoryChange}
          onSearchCompanyChange={onSearchCompanyChange}
          onClose={onClose}
        />
      );

      expect(document.body.classList.contains('no-scroll')).toBe(true);
    });

    it('removes the no-scroll class from the body when isOpen is false', () => {
      render(
        <Default
          isOpen={false}
          onSearchCategoryChange={onSearchCategoryChange}
          onSearchCompanyChange={onSearchCompanyChange}
          onClose={onClose}
        />
      );

      expect(document.body.classList.contains('no-scroll')).toBe(false);
    });

    it('removes the no-scroll class from the body when the component unmounts', () => {
      const { unmount } = render(
        <Default
          isOpen={true}
          onSearchCategoryChange={onSearchCategoryChange}
          onSearchCompanyChange={onSearchCompanyChange}
          onClose={onClose}
        />
      );

      expect(document.body.classList.contains('no-scroll')).toBe(true);

      unmount();

      expect(document.body.classList.contains('no-scroll')).toBe(false);
    });
  });
});
