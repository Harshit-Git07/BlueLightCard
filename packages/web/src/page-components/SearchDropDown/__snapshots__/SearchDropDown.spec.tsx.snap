// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchDropDown Component Basic Rendering renders the dropdown correctly 1`] = `
<div>
  <div
    class="w-full"
  >
    <div
      aria-hidden="true"
      class="w-full h-[100vh] bg-black/50 absolute backdrop-blur-sm"
      data-testid="search-dropdown-overlay"
      id="search-dropdown-overlay"
    />
    <dialog
      class="w-full h-auto max-h-[80vh] bg-colour-surface dark:bg-colour-surface-dark block absolute z-10 overflow-y-auto"
    >
      <div
        class="mx-auto tablet:pt-[54px] pb-[100px] max-w-[1408px]"
      >
        <div
          class="mb-[22px] px-4"
        >
          <div
            class="w-full"
          >
            <p
              class="py-1.5 mb-2 font-typography-title-large font-typography-title-large-weight text-typography-title-large leading-typography-title-large tracking-typography-title-large text-colour-onSurface-light dark:text-colour-onSurface-dark"
            >
              Browse categories
            </p>
            <div
              class="flex gap-[15px] flex-wrap"
            >
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="0"
              >
                Children and toys
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="1"
              >
                Days out
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="2"
              >
                Electrical and phones
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="3"
              >
                Entertainment
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="4"
              >
                Fashion
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="5"
              >
                Featured
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="6"
              >
                Financial and insurance
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="7"
              >
                Food and drink
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="8"
              >
                Gifts
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="9"
              >
                Health and beauty
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="10"
              >
                Holidays and travel
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="11"
              >
                Home and garden
              </button>
              <button
                class="font-pill-label-font text-pill-label-font font-pill-label-font-weight tracking-pill-label-font leading-pill-label-font rounded-full px-3 whitespace-nowrap min-w-[54px] py-2 bg-pill-default-bg-colour-light dark:bg-pill-default-bg-colour-dark text-pill-default-label-colour-light dark:text-pill-default-label-colour-dark"
                tabindex="12"
              >
                Motoring
              </button>
            </div>
          </div>
        </div>
        <div
          class="px-4"
        >
          <p
            class="mb-2 font-typography-title-large font-typography-title-large-weight text-typography-title-large leading-typography-title-large tracking-typography-title-large text-colour-onSurface-light dark:text-colour-onSurface-dark"
          >
            Browse companies
          </p>
          <div
            class="group relative inline-block w-full"
          >
            <div
              class="relative"
            >
              <input
                aria-controls="dropdown-listbox-:r0:"
                aria-expanded="false"
                aria-label="Search for a company"
                autocomplete="off"
                class="z-20 w-full h-[50px] px-[16px] text-left pb-2 pt-5 peer font-typography-body font-typography-body-weight text-typography-body leading-typography-body tracking-typography-body text-colour-onSurface dark:text-colour-onSurface-dark cursor-pointer bg-transparent rounded-md border outline-none border-colour-onSurface-outline dark:border-colour-onSurface-outline-dark"
                data-testid="combobox"
                id=":r0:"
                role="combobox"
                spellcheck="false"
                tabindex="0"
                type="text"
                value=""
              />
              <label
                aria-disabled="false"
                aria-hidden="false"
                class="left-4 -translate-y-1/2 absolute transition-all pointer-events-none top-[25px] peer-focus:top-4 font-typography-body font-typography-body-weight text-typography-body leading-typography-body tracking-typography-body peer-focus:font-typography-body-light peer-focus:font-typography-body-light-weight peer-focus:text-typography-body-light peer-focus:leading-typography-body-light peer-focus:tracking-typography-body-light peer-focus:text-xs text-colour-onSurface-subtle dark:text-colour-onSurface-subtle-dark"
                for=":r0:"
                id=":r0:-placeholder"
              >
                Search for a company
              </label>
              <button
                aria-label="Open dropdown"
                class="absolute top-1/2 right-4 -translate-y-1/2 pointer-events-none false"
                title="Open dropdown"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="svg-inline--fa fa-chevron-down w-[16px] h-[16px] text-colour-onSurface dark:text-colour-onSurface-dark"
                  data-icon="chevron-down"
                  data-prefix="fas"
                  focusable="false"
                  role="img"
                  viewBox="0 0 512 512"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M233.4 406.6c12.5 12.5 32.8 12.5 45.3 0l192-192c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0L256 338.7 86.6 169.4c-12.5-12.5-32.8-12.5-45.3 0s-12.5 32.8 0 45.3l192 192z"
                    fill="currentColor"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>
    </dialog>
  </div>
</div>
`;
